Add additional str accessor methods for DataArray

toddrjen opened the pull request:
This implements the following additional string accessor methods, based loosely on the versions in pandas:

#### One-to-one

- [x] casefold(self)
- [x] normalize(self, form)

#### One-to-many

- [x] extract(self, pat[, flags, expand])
- [x] extractall(self, pat[, flags])
- [x] findall(self, pat[, flags])
- [x] get_dummies(self[, sep])
- [x] partition(self[, sep, expand])
- [x] rpartition(self[, sep, expand])
- [x] rsplit(self[, pat, n, expand])
- [x] split(self[, pat, n, expand])

#### Many-to-one

- [x] cat(self[, others, sep, na_rep, join])
- [x] join(self, sep)

#### Operators

- [x] `+`
- [x] `*`
- [x] `%`

#### Other

- [x] Allow vectorized arguments.

 - [x] Closes #3940
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
 - [x] New functions/methods are listed in `api.rst`

max-sixty reviewed (commented):
Hi @toddrjen — this is very impressive and quite a lot! Let me / others take some time this week to go through it more.

I took a look though `split` & `rsplit` as a sample, which look excellent. Thanks for the docstrings — those make them very clear.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -76,17 +80,46 @@ class StringAccessor:
+        g = np.vectorize(f, otypes=[otype], signature=signature)
+        return apply_ufunc(g, obj, dask="parallelized", output_dtypes=[dtype], **kwargs)
+
+    def _check_is_compiled_re(self, pat):
```
I did that originally, but I thought this made it clearer exactly what is going on.  I could change it back if you prefer.

max-sixty replied:
I don't have a strong view, we can see what others think.

toddrjen replied:
Thinking more I don't think it makes much sense anymore so I went ahead and removed it.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -76,17 +80,46 @@ class StringAccessor:
 
-        g = np.vectorize(f, otypes=[dtype])
-        return apply_ufunc(g, self._obj, dask="parallelized", output_dtypes=[dtype])
+        g = np.vectorize(f, otypes=[otype], signature=signature)
+        return apply_ufunc(g, obj, dask="parallelized", output_dtypes=[dtype], **kwargs)
```
```suggestion
        return apply_ufunc(g, obj, dask="parallelized", output_dtypes=[dtype], vectorize=True, **kwargs)
```

`apply_ufunc` has a `vectorize` keyword. It also automatically infers the signature (given the correct `input_`- and `output_core_dims`.

toddrjen replied:
Fixed, see below.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -76,17 +80,46 @@ class StringAccessor:
         self._obj = obj
 
-    def _apply(self, f, dtype=None):
+    def _apply(self, f, obj=None, dtype=None, otype=None, signature=None, **kwargs):
```
```suggestion
    def _apply(self, f, obj=None, dtype=None, otype=None, **kwargs):
```
see below

toddrjen replied:
Fixed, see below.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -76,17 +80,46 @@ class StringAccessor:
+            obj = self._obj
         if dtype is None:
-            dtype = self._obj.dtype
+            dtype = obj.dtype
+        if otype is None:
+            otype = dtype
```
What is the difference between `dtype` and `otype` should they not be the same?

mathause replied:
Ok I see below why but what is `result.dtype` when `dtype` and `otype` differ? Consider the following example:

```python
import numpy as np
import xarray as xr

da = xr.DataArray(["a", "b", "cd"])
print(da.dtype)

g = np.vectorize(lambda x: x.upper(), otypes=[np.object_])
xr.apply_ufunc(g, da, output_dtypes=[da.dtype]).dtype # -> is object
```

so it seems there is no point in making a difference between `dtype` and `otype` - am I wrong?

toddrjen replied:
I didn't realize that `assert_equal` wasn't comparing dtypes.  So you are right, this was returning a different dtype.  I fixed this and simplified the code.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -129,7 +166,12 @@ def f(x):
+        start: int = None,
+        stop: int = None,
+        step: int = None,
+    ) -> Any:
```
```suggestion
    ) -> "DataArray":
```
isn't that always a DataArray? (maybe this does not hold for all functions but most I think)

toddrjen replied:
In our usage, but there is nothing stopping someone from using this with another class with a similar API.

toddrjen replied:
The return type for `apply_ufunc` is also `Any`, so I think it would make sense to use the same thing.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -574,15 +706,31 @@ def match(self, pat, case=True, flags=0):
         """
-        if not case:
-            flags |= re.IGNORECASE
+        is_compiled_re = self._check_is_compiled_re(pat)
```
Isn't this code block equal to `_re_compile`?

toddrjen replied:
Yes, I missed that one.  Fixed.

mathause reviewed (commented):
I had a quick look and added some suggestions/ questions. I'll need to give it a more thorough review but looks good so far.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        def _get_res_multi(val, pat=pat):
+            match = pat.search(val)
+            if match is None:
+                return np.array([""], val.dtype)
```
```suggestion
                return np.array([""] * n_groups, val.dtype)
```
Does this not need to have the same size as there are numbers of groups?

toddrjen replied:
Yes, I will add a unit test for this.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+    def extractall(
+        self,
+        pat: Any,
+        group_dim: Hashable,
+        match_dim: Hashable,
```
```suggestion
        group_dim: Hashable = "group",
        match_dim: Hashable = "match",
```
I'd vote for giving sensible defaults here.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        Extract matches
+
+        >>> value.str.extractall(r"(\\w+)_Xy_(\\d*)", group_dim="XX", match_dim="YY")
```
```suggestion
        >>> value.str.extractall(r"(\\w+)_Xy_(\\d*)", group_dim="group", match_dim="match")
```
I'd find that easier to understand.

toddrjen replied:
I fixed up this issue throughout the examples.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+                output_sizes={dim: pat.groups},
+            ).astype(self._obj.dtype.kind)
+
+    def extractall(
```
Just for my understanding - this is relevant if there is more than one match per group?

toddrjen replied:
Correct.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        _get_count = lambda x: len(pat.findall(x))
+        maxcount = self._apply(_get_count, dtype=np.int_).max().data.tolist()
+
+        def _get_res(val, pat=pat, num=maxcount, dtype=self._obj.dtype):
```
```suggestion
        def _get_res(val):
```
I think you can remove the args - `_get_res` is in the namespace of `extractall`

toddrjen replied:
This guarantees that the names are bound with the correct values and not with other values that may be set later.  It prevents potential gotchas if variables are modified.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        def _get_res(val, pat=pat, num=maxcount, dtype=self._obj.dtype):
+            matches = pat.findall(val)
+            res = np.zeros([maxcount, pat.groups], dtype)
```
Else this would have to be `num`

toddrjen replied:
Fixed.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            output_sizes={group_dim: pat.groups, match_dim: maxcount},
+        ).astype(self._obj.dtype.kind)
+
+    def findall(
```
I am a bit skeptical how useful a DataArray of lists is.

toddrjen replied:
It would be a step in further processing I would think.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        Parameters
+        ----------
+        pat : str or re.Pattern
+            A string contain a regular expression or
+            a compiled regular expression object.
```
```suggestion
            A string containing a regular expression or a compiled regular
            expression object.
```

toddrjen replied:
Fixed.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            a compiled regular expression object.
+        case : bool, default: True
+            If True, case sensitive.
+            Cannoy be set if `pat` is a compiled regular expression.
```
```suggestion
            Cannot be set if `pat` is a compiled regular expression.
```

toddrjen replied:
Fixed

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        Returns
+        -------
+        extracted : same type as values or object array
```
That will have `dtype=object` right? Also the ones below.

toddrjen replied:
Fixed.  This is the only one that only returns object arrays, though

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+    def partition(
+        self,
+        dim: Optional[Hashable],
```
```suggestion
        dim: Optional[Hashable] = "parts",
```
I would also give a reasonable default here. I feel a DataArray of lists is not very useful.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+    def rpartition(
+        self,
+        dim: Optional[Hashable],
```
```suggestion
        dim: Optional[Hashable] = "parts",
```

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+    def split(
+        self,
+        dim: Optional[Hashable],
```
```suggestion
        dim: Optional[Hashable] = "parts",
```

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            f = lambda x: x.split(sep, maxsplit)
+            return self._apply(f, dtype=np.object_)
+
+        if not self._obj.size:
+            return self._obj.copy().expand_dims({dim: 0}, -1)
```
What are you doing here?

toddrjen replied:
Working around an issue in `_apply`.  I added a comment explaining this.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+    def rsplit(
+        self,
+        dim: Optional[Hashable],
```
```suggestion
        dim: Optional[Hashable] = "parts",
```

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            output_sizes={dim: maxsplit},
+        ).astype(self._obj.dtype.kind)
+
+    def rsplit(
```
Could you combine this with `split`

```python
def _split(...):
    ...

def split(...):
    def splitter(sep, maxsplit):
        return x.split(sep, maxsplit)
    return _split(..., splitter)

def rsplit(...):
    def splitter(sep, maxsplit):
        return x.rsplit(sep, maxsplit)
    return _split(..., splitter)

```
The same with `partition` and `rpartition`.

toddrjen replied:
Done.  I couldn't combine split/rsplit with partition/rpartition, though, without making things really complicated

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            If `None`, place the results as list elements in an object DataArray
+        sep : str, default is split on any whitespace.
+            String to split on.
+        maxsplit : int, defualt -1 (all)
```
```suggestion
        maxsplit : int, default -1 (all)
```

toddrjen replied:
Fixed

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            String to split on.
+        maxsplit : int, defualt -1 (all)
+            Limit number of splits in output, starting from the end.
+            -1 will be interpreted as return all splits.
```
```suggestion
            -1 will return all splits.
```

toddrjen replied:
This was taken from pandas, but your version is probably better.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        maxsplit : int, defualt -1 (all)
+            Limit number of splits in output, starting from the end.
+            -1 will be interpreted as return all splits.
+            The final number of split values may be less than this if no
```
```suggestion
            The final number of split values may be less than this if there are no
```

toddrjen replied:
Fixed

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        dim : Hashable or `None`
+            Name for the dimension to place the results in.
+            If `None`, place the results as list elements in an object DataArray
+        sep : str, default is split on any whitespace.
+            String to split on.
```
How does that work if the default is `None`? Could you set this to `"\w"`?

toddrjen replied:
It is handled by the `str.split` method.  The unit tests check this.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+    def get_dummies(
+        self,
+        dim: Hashable,
+        sep: str = "|",
```
Why a different default here?

toddrjen replied:
This is taken from `pandas`.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+    def get_dummies(
+        self,
+        dim: Hashable,
```
```suggestion
        dim: Hashable = "groups",
```
Or `"dummies"`?

toddrjen replied:
And what if there is already a dimension named "groups"?  This would be a serious problem if someone does operations multiple times.  

Also, I checked how xarray normally handles automatic dimension names, and except for the constructor it generally doesn't.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        f = lambda x: np.array([val in x for val in vals], dtype=np.bool_)
+        res = self._apply(
+            f,
+            obj=setarr,
```
As far as I can see this is the only place you use `obj` - could this be refactored?

toddrjen replied:
Do you have a way to do that which wouldn't require considerably more code and complexity?

mathause reviewed (commented):
I went over the code and added a lot of comments. But mostly because it's so long. As already mentioned looks good overall.

My main points:
- I'd set a default for the name of new dimensions e.g. `group_dim: Hashable = "group"`. I think that's a good choice in most cases.
- I'd recommend to refactor the complicated `XX` and `rXX` methods, e.g. `split` and `rsplit` (see inline comments).

Could you add the methods to the docs:
- https://github.com/pydata/xarray/blob/master/doc/api.rst#string-manipulation
- https://github.com/pydata/xarray/blob/master/doc/api-hidden.rst

I have not checked the tests yet.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -239,6 +301,44 @@ def upper(self):
+        Convert strings in the array to be casefolded.
+
+        Casefolding is similar to converting to lowercase,
+        but removes all case distinctions.
+        This is important in some languages that have more complicated
```
No, this is meant to be a single paragraph.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -76,17 +80,61 @@ class StringAccessor:
+        if not case:
+            flags |= re.IGNORECASE
+
+        pat = self._obj.dtype.type(pat)
```
Done

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -239,6 +301,44 @@ def upper(self):
+
+        Parameters
+        ----------
+        side : {"NFC", "NFKC", "NFD", and "NFKD"}
```
Done

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -348,8 +453,11 @@ def count(self, pat, flags=0):
-        pat : str
-            Valid regular expression.
+        pat : str or re.Pattern
+            A string contain a regular expression or
+            a compiled regular expression object.
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -338,7 +438,12 @@ def isupper(self):
+    def count(
+        self,
+        pat: Any,
+        case: bool = True,
+        flags: int = 0,
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -348,8 +453,11 @@ def count(self, pat, flags=0):
+            If True, case sensitive.
         flags : int, default: 0
             Flags for the `re` module. Use 0 for no flags. For a complete list,
             `see here <https://docs.python.org/3/howto/regex.html#compilation-flags>`_.
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            resulting DataArray will have the same size as the original.
+        case : bool, default: True
+            If True, case sensitive.
+            Cannoy be set if `pat` is a compiled regular expression.
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        Raises
+        ------
+        ValueError : `pat` has no capture groups.
+        ValueError : `dim` is `None` and there is more than one capture group.
+        ValueError : `case` is set when `pat` is a compiled regular expression.
+        KeyError : The given dimension is already present in the DataArray.
```
I got the format wrong.  I have changed this to use the correct Raises block formatting.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            raise ValueError("No capture groups found in pattern.")
+
+        if dim is None and pat.groups != 1:
+            raise ValueError("dim must be specified if more than one capture group.")
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            match = pat.search(val)
+            if match is None:
+                return ""
+            res = match.group(1)
```
`0` is the whole match, `1` is the first capture group.  From the [documentation](https://docs.python.org/3/library/re.html#re.Match.group):

> If a groupN argument is zero, the corresponding return value is the entire matching string; if it is in the inclusive range [1..99], it is the string matching the corresponding parenthesized group.

The unit tests check this.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -865,49 +1061,838 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            return np.array(match, val.dtype)
+
+        if dim is None:
+            return self._apply(_get_res_single)
```
No, if `dim` is `None` then the output has the same shape as the input.  `np.object_` is only needed if the output has more dimensions than the output.

toddrjen commented:
> * I'd set a default for the name of new dimensions e.g. `group_dim: Hashable = "group"`. I think that's a good choice in most cases.

I thought about doing this at first.  However, this could lead to conflicts if the DataArray already has a dimension with that name, which would be a particular problem if people chained together multiple such operations.  So I checked what default name xarray uses elsewhere, and it doesn't seem to use default names for the most part (the main exception being DataArray creation).  So I think that, in order to avoid unexpected behavior, and to keep consistency, not automatically choosing a name is a better option.

mathause commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,10 +322,893 @@ def test_replace_literal(dtype):
+
+def test_extract_single_nocase(dtype):
+    pat_str = r"(\w+)_Xy_\d*"
+    conv = {np.unicode_: str, np.bytes_: lambda x: bytes(x, encoding="UTF-8")}[dtype]
+    pat_re = re.compile(conv(pat_str), flags=re.IGNORECASE)
```
I think this should be easier to understand:
```suggestion
    pat_str = pat_str if dtype is np.str_ else bytes(pat_str, encoding="UTF-8")
    pat_re = re.compile(pat_str, flags=re.IGNORECASE)
```
---
(Alternatively:



```suggestion
    pat_str = {np.unicode_: pat_str, np.bytes_: bytes(pat_str, encoding="UTF-8")}[dtype]
    pat_re = re.compile(pat_str, flags=re.IGNORECASE)
```
)

toddrjen replied:
Fixed

mathause commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,10 +322,893 @@ def test_replace_literal(dtype):
+        dims=["X", "Y"],
+    ).astype(dtype)
+
+    targ = xr.DataArray(
```
```suggestion
    expected = xr.DataArray(
```

toddrjen replied:
Fixed everywhere

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+                return ""
+            res = match.group(1)
+            if res is None:
+                res = ""
```
This line is not covered with a test.

toddrjen replied:
Fixed.

mathause commented on xarray/core/accessor_str.py:
```diff
@@ -537,52 +706,66 @@ def contains(self, pat, case=True, flags=0, regex=True):
-            regex = re.compile(pat, flags=flags)
+        is_compiled_re = isinstance(pat, self._pattern_type)
+        if is_compiled_re and not regex:
+            raise ValueError(
```
This line is not covered with a test.

toddrjen replied:
Fixed.

mathause commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,10 +322,893 @@ def test_replace_literal(dtype):
+        dims=["X", "Y"],
+    ).astype(dtype)
+
+    targ = [[["a"], [], ["abc"]], [["abcd"], [], ["abcdef"]]]
```
```suggestion
    expected = [[["a"], [], ["abc"]], [["abcd"], [], ["abcdef"]]]
```

mathause reviewed (commented):
> However, this could lead to conflicts if the DataArray already has a dimension with that name, which would be a particular problem if people chained together multiple such operations.

That should raise a KeyError, no?

> So I checked what default name xarray uses elsewhere, and it doesn't seem to use default names for the most part (the main exception being DataArray creation)

Yes this is true. I'd still prefer defaults but I leave it up to you.

---
I had a short look at the tests (could not go through them all right now). Again, what I saw looks good. Some things that I noticed:

- The test coverage is very good.
- Some of the tests could probably be simplified, to make them easier to read. E.g. when you try to raise an error.
- we usually add a `match` to the `pytest.raises`. This also helps to understand what you are testing.
- `assert_equal` should raise an error if the dtype does not match, so you should not need to add all the `assert result.dtype == expected.dtype`. Consider

```python
import numpy as np
import xarray as xr

a = xr.DataArray(np.array("a", dtype=np.str_))
b = xr.DataArray(np.array("a", dtype=np.bytes_))

xr.testing.assert_equal(a, b)
```

mathause commented:
Just wanted to let you know that we are definitively interested in this contribution! I'd start by removing the dtype assertions again, that makes the diff of the tests much smaller and more digestible. Unless there is a reason for them? Does this not work correctly in `assert_equal`?

toddrjen commented:
@mathause 

Sorry for the delay, I have been swamped at work.  I probably won't have any time to work on this before Christmas.

I have finished implementing the `cat` and `join` methods, and I implemented `+`, `*`, and `%` operator support.  

I am currently working on improving the vectorization of some of the functions.  The idea is that some arguments, like for example the regular expression pattern or the number of repetitions in rep, will be able to be given an array-like, with the dimensions being broadcast against the original DataArray.  

This can be useful, for example, if a DataArray combines data of different formats along a dimension (ideally this wouldn't be the case but people don't always have that much control over the data they get).  Or it could be used to create an ASCII bar chart where the number of symbols is equal to the value in an array element.

> > However, this could lead to conflicts if the DataArray already has a dimension with that name, which would be a particular problem if people chained together multiple such operations.
> 
> That should raise a KeyError, no?

Yes, but I think it would be strange if using the default parameters once works fine, but using them twice or more in a row somehow returns an exception.  I think the defaults should either work generally or not be defaults at all.  That is just my opinion.  More fundamentally, it is just inconsistent with how xarray works elsewhere and so I think it would be unexpected.

> * Some of the tests could probably be simplified, to make them easier to read. E.g. when you try to raise an error.

Please point out the specific cases if you haven't already done so.

> * we usually add a `match` to the `pytest.raises`. This also helps to understand what you are testing.

I will add this.

> * `assert_equal` should raise an error if the dtype does not match, so you should not need to add all the `assert result.dtype == expected.dtype`. 

It doesn't work with an object dtype:

```python
>>> import numpy as np 
>>> import xarray as xr 
>>>   
>>> a = xr.DataArray(np.array("a", dtype=np.str_)) 
>>> b = a.astype(np.object_)                     
>>> a.dtype == b.dtype                                                                                                                                                                                                                        
False
>>> a.equals(b)
True
>>> xr.testing.assert_equal(a, b)
```

This does not raise an exception on my machine at least.  I ran into several cases where I was incorrectly getting object dtypes and the tests weren't catching it, hence the dtype checks.

mathause commented:
> Sorry for the delay, I have been swamped at work. I probably won't have any time to work on this before Christmas.

No probem at all, just wanted to check in.

Yes you are correct concerning the dtype. This comes back to numpy, where the following returns true

```python
np.array("a") == np.array("a", dtype=object)
```
I wonder if that's the right choice... Thus, you are right to change the tests.

toddrjen commented:
@mathause One possibility might be to make `xr.testing.assert_identical` match dtypes.  I can see different dtypes being "equal", but not "identical".

mathause commented:
I opened an issue regarding the dtypes check. Let's see what the others think.

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -1,3 +1,5 @@
+# -*- coding: utf-8 -*-
```
It is used by a variety of tools and text editors besides Python, so I would prefer to keep it in.

keewis replied:
the magic comment does not really hurt since `utf-8` is the default source code encoding for python 3. However, we don't have editor specific comments ([PEP 263](https://www.python.org/dev/peps/pep-0263/) calls the format you used the "emacs style") anywhere else so for consistency it might be better to remove it.

andersy005 replied:
Do we still want this to be addressed before merging?

toddrjen replied:
Fixed.

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -61,60 +63,143 @@ def test_dask():
-    assert_equal(da.str.swapcase(), swapped)
-    assert_equal(da.str.title(), titled)
-    assert_equal(da.str.upper(), uppered)
+def test_case_bytes(dtype):
```
Fixed

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -61,60 +63,143 @@ def test_dask():
+    assert res_titled.dtype == exp_titled.dtype
+    assert res_uppered.dtype == exp_uppered.dtype
+
+    assert_equal(value.str.capitalize(), exp_capitalized)
```
Fixed

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,10 +322,893 @@ def test_replace_literal(dtype):
+        dims=["X", "Y"],
+    ).astype(dtype)
+
+    with pytest.raises(ValueError):
```
Fixed here and elsewhere

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,10 +322,893 @@ def test_replace_literal(dtype):
         values.str.replace(compiled_pat, "", regex=False)
 
 
+def test_extract_extractall_findall_empty_raises(dtype):
```
Fixed.

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        dim : Hashable or `None`
+            Name for the dimension to place the results in.
+            If `None`, place the results as list elements in an object DataArray
+        sep : str, default is split on any whitespace.
+            String to split on.
```
This would be interpreted as a (broken) link to `is split on any whitespace` by `napoleon`, so I would move it to the parameter description:
```suggestion
        sep : str, default: None
            String to split on. If ``None``, split on any whitespace.
```

toddrjen replied:
Fixed

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        Parameters
+        ----------
+        dim : Hashable or `None`
```
```suggestion
        dim : hashable or None
```

toddrjen replied:
Fixed

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -239,7 +320,45 @@ def upper(self):
+
+        Parameters
+        ----------
+        form : {"NFC", "NFKC", "NFD", and "NFKD"}
```
```suggestion
        form : {"NFC", "NFKC", "NFD", "NFKD"}
```

toddrjen replied:
Fixed

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        pat : str or re.Pattern
+            A string containing a regular expression or a compiled regular
+            expression object.
+        dim : hashable or `None`
```
```suggestion
        dim : hashable or None
```

toddrjen replied:
Fixed

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        Parameters
+        ----------
+        dim : Hashable or `None`
```
```suggestion
        dim : hashable or None
```

toddrjen replied:
Fixed

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        Parameters
+        ----------
+        dim : Hashable or `None`
```
```suggestion
        dim : hashable or None
```

toddrjen replied:
Fixed.

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        dim : Hashable or `None`
+            Name for the dimension to place the results in.
+            If `None`, place the results as list elements in an object DataArray
+        sep : str, default is split on any whitespace.
```
```suggestion
        sep : str, default: None
            String to split on. If ``None``, split on any whitespace.
```

toddrjen replied:
Fixed.

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            If `None`, place the results as list elements in an object DataArray
+        sep : str, default is split on any whitespace.
+            String to split on.
+        maxsplit : int, default -1 (all)
```
```suggestion
        maxsplit : int, default: -1
```

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        Parameters
+        ----------
+        dim : Hashable
```
```suggestion
        dim : hashable
```

toddrjen replied:
Fixed

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        """
+        # _apply breaks on an empty array in this case
+        if not self._obj.size:
+            return self._obj.copy().expand_dims({dim: 0}, -1)
```
```suggestion
            return self._obj.copy().expand_dims({dim: 0}, axis=-1)
```

toddrjen replied:
Fixed.

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        # _apply breaks on an empty array in this case
+        if not self._obj.size:
+            return self._obj.copy().expand_dims({dim: 0}, -1)
```
```suggestion
            return self._obj.copy().expand_dims({dim: 0}, axis=-1)
```

toddrjen replied:
Fixed

keewis commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            return self._obj.copy().expand_dims({dim: 0}, -1)
+
+        f_count = lambda x: max(len(func(x, sep, maxsplit)), 1)
+        maxsplit = self._apply(f_count, dtype=np.int_).max().data.tolist() - 1
```
is the `tolist` necessary here? As far as I can tell a bare `max` will always return a `numpy` scalar (wrapped by a `xarray` object). If it is, using `.item()` might be cleaner.

toddrjen replied:
That seems to work.

keewis reviewed (commented):
I've got a few more suggestions, most of them regarding the docstrings.

Could you add the new methods to the `DataArray.str` section in `api.rst`?

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -76,19 +80,79 @@ class StringAccessor:
     """
 
     __slots__ = ("_obj",)
+    _pattern_type = type(re.compile(""))
```
This is [new in Python 3.7](https://docs.python.org/3/whatsnew/changelog.html#id241). Python 3.6 is still supported as far as I can tell.

keewis replied:
thanks, I didn't notice that. We're currently removing support for Python 3.6 (see #4720), but I think it is fine to keep it as is for now.

toddrjen replied:
To keep the unit tests working I will keep it as-is until 3.6 is fully dropped.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        Parameters
+        ----------
+        dim : Hashable or `None`
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -239,7 +320,45 @@ def upper(self):
+        -------
+        casefolded : same type as values
+        """
+        return self._apply(lambda x: x.casefold())
```
Yes, I think that version of the error message is clearer.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        pat : str or re.Pattern
+            A string containing a regular expression or a compiled regular
+            expression object.
+        group_dim: hashable
```
Fixed.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        group_dim: hashable
+            Name of the new dimensions corresponding to the capture groups.
+            This dimension is added to the new DataArray first.
+        match_dim: hashable
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        dim : Hashable or `None`
+            Name for the dimension to place the 3 elements in.
+            If `None`, place the results as list elements in an object DataArray
+        sep : str, default `" "`
```
Fixed.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        dim : Hashable or `None`
+            Name for the dimension to place the 3 elements in.
+            If `None`, place the results as list elements in an object DataArray
+        sep : str, default `" "`
```
Fixed.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+        ----------
+        dim : Hashable
+            Name for the dimension to place the results in.
+        sep : str, default `"|"`.
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1085,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            If `None`, place the results as list elements in an object DataArray
+        sep : str, default is split on any whitespace.
+            String to split on.
+        maxsplit : int, default -1 (all)
```
Fixed

toddrjen commented:
@keewis Thanks for the suggestions.  I will add everything to the relevant documentation when I have everything completed and the changes are agreed upon.

keewis commented:
`api.rst` contains a structured enumeration of methods `autosummary` should generate documentation pages for, so there's not much to do, and it would allow us to check the rendered version of the docstrings.

toddrjen commented:
The latest version I just pushed should have the requested changes.  It also has `cat`, `join`, `+`, `*`, `%`.  I have also implemented broadcasting for many (but not all) of the functions I plan to implement it for so you can see some examples of how it works.

mathause reviewed (approved):
I had another look and have some more comments. However, I think none of them are blocking - so I'd say you implement the suggestions you deem worthy & tell me when you are happy. Then we give the others another day or two to comment & then I'll merge.

---

- I think code blocks in the docstring need a double backticks (\`). Would be nice if you can fix those. 
- I think is more usual to only have one space after a full stop in our docs.
- broadcast
  - Usually xarray uses `join="outer"` when broadcasting. Here `join="exact"` is used. This makes total sense as missing values are not handled. But this might be surprising for users.
  - I think the formulation `"If array-like, it is broadcast."` could be improved.
  - How about `Array-like input is broadcast using join="exact"`?
- In your examples you directly showcase the full functionality. E.g. in `cat` you have two arrays with different dimensions, 3 scalar inputs and the separator is along a third dimension all in one example. This can be quite difficult to wrap my head around. I would probably have made 3 examples out of this (1D array + 1 or 2 scalars + scalar sep; 2 x 1D arrays; 1 D array + 1D sep), making it easier to digest.
- The same holds for the tests. 
- These last two points do not mean you have to change them (all), and I appreciate how thorough you were, thinking of all the corner cases. However, from a usability and maintainability perspective less can also sometimes be more.

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,479 +383,2704 @@ def test_replace_literal(dtype):
+
+
+def test_format_scalar():
+    dtype = np.unicode_
```
In some situations object arrays could be created, so I want to force it to be a string dtype to be certain. However, it doesn't need to be a variable, so I will just apply the dtype directly.

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,479 +383,2704 @@ def test_replace_literal(dtype):
+    assert_equal(res, expected)
+
+
+def test_format_broadcast(dtype):
```
Fixed

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,479 +383,2704 @@ def test_replace_literal(dtype):
+        dims=["X"],
+    ).astype(dtype)
+
+    pos0 = 1
```
It is unclear to me at first glance what that would do, and I prefer to keep the tests as straightforward as possible. So I will keep it as-is.

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,479 +383,2704 @@ def test_replace_literal(dtype):
+    res_blank_y = values.str.join(dim="Y")
+
+    res_space_x = values.str.join(dim="X", sep=" ")
+    res_space_y = values.str.join(sep=" ", dim="Y")
```
Fixed.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -120,76 +284,331 @@ def get(self, i, default=""):
+
+        This is equivalent to calling `str.format` on every element of the
+        DataArray.  The replacement values can either be a string-like
+        scalar or an array-like of string-like values.  If array-like,
```
Fixed all these as far as I can tell.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -73,22 +77,146 @@ class StringAccessor:
         array([4, 4, 2, 2, 5])
         Dimensions without coordinates: dim_0
 
+    It also implements `+`, `*`, and `%`, which operate as elementwise
+    versions of the corresponding `str` methods.  These will automatically
```
Fixed.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -73,22 +77,146 @@ class StringAccessor:
+        Dimensions without coordinates: X, Y, Z
+
+    .. note::
+        When using `%` formatting with a dict, the values are always used as a
+        single value, they aren't applied elementwise.
```
Fixed.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -120,76 +284,331 @@ def get(self, i, default=""):
+            dtype=self._obj.dtype.kind,
+        )
+
+    def join(
```
Done.

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,479 +383,2704 @@ def test_replace_literal(dtype):
+    assert_equal(result, exp)
+
+    for start, stop, step in [(0, 3, -1), (None, None, -1), (3, 10, 2), (3, 0, -1)]:
+        try:
```
It doesn't hurt. I think it is still good to test this.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -120,76 +284,331 @@ def get(self, i, default=""):
+               ['abcd--abcdef', 'abcd__abcdef']], dtype='<U12')
+        Dimensions without coordinates: X, ZZ
+
+
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -239,7 +658,45 @@ def upper(self):
+        Returns
+        -------
+        normalized : same type as values
+
+
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -338,159 +795,226 @@ def isupper(self):
         ----------
         pat : str
             Character sequence. Regular expressions are not accepted.
+             If array-like, it is broadcast.
```
I think I fixed this

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -338,159 +795,226 @@ def isupper(self):
         """
         Test if the end of each string in the array matches a pattern.
 
+        The pattern can either be a `str` or array-like of `str`.
```
I think I fixed this

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -338,159 +795,226 @@ def isupper(self):
-        width : int
+        width : int or array-like of int
             Minimum width of resulting string; additional characters will be
             filled with character defined in `fillchar`.
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -338,159 +795,226 @@ def isupper(self):
+        fillchar = self._stringify(fillchar)
+
+        def overfunc(x, iwidth, ifillchar):
+            if len(ifillchar) != 1:
+                raise TypeError("fillchar must be a character, not str")
```
Fixed

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,479 +383,2704 @@ def test_replace_literal(dtype):
+def test_find(dtype):
+    values = xr.DataArray(["ABCDEFG", "BCDEFEF", "DEFGHIJEF", "EFGHEF", "XXX"])
+    values = values.astype(dtype)
+    result = values.str.find("EF")
```
I think this line is no longer in the code.

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,479 +383,2704 @@ def test_replace_literal(dtype):
+    assert empty_bool.dtype == empty.str.istitle().dtype
+    assert empty_bool.dtype == empty.str.isupper().dtype
+    assert empty_bytes.dtype.kind == empty.str.encode("ascii").dtype.kind
+    assert empty_int.dtype.kind == empty.str.count("a").dtype.kind
+    assert empty_int.dtype.kind == empty.str.find("a").dtype.kind
+    assert empty_int.dtype.kind == empty.str.len().dtype.kind
+    assert empty_int.dtype.kind == empty.str.rfind("a").dtype.kind
```
empty strings can have seemingly random string lengths so this doesn't work.

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -224,479 +383,2704 @@ def test_replace_literal(dtype):
+    assert empty_str.dtype.kind == empty.str.strip().dtype.kind
+    assert empty_str.dtype.kind == empty.str.swapcase().dtype.kind
+    assert empty_str.dtype.kind == empty.str.title().dtype.kind
+    assert empty_str.dtype.kind == empty.str.upper().dtype.kind
```
Yes, I don't know how to predict the dtypes of arrays of empty strings.

toddrjen commented on xarray/tests/test_accessor_str.py:
```diff
@@ -44,7 +46,7 @@
 import xarray as xr
 
-from . import assert_equal, requires_dask
+from . import assert_equal, assert_identical, requires_dask
```
I'll keep this until there is a proper solution in xarray.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1452,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        Extract matches
+
+        >>> value.str.findall(r"(\\w+)_Xy_(\\d*)")
```
I think the `\\` turns to `\` when it is rendered.

keewis replied:
the docstring will evaluate the escapes on import and the doctest would evaluate the result, which would require `\\\\w` for normal strings. This seems a bit confusing and makes copying the example difficult, though, so maybe you should mark the whole docstring as "raw", too: `r"""docstring"""`, which would allow us to have the docstring and the rendered result look the same.

toddrjen replied:
Okay, done.

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1452,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+
+        if group_dim == match_dim:
+            raise KeyError(
+                f"Group dimension {group_dim} is the same as match dimension {match_dim}."
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1452,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            raise KeyError(f"Group dimension {group_dim} already present in DataArray.")
+
+        if match_dim in self._obj.dims:
+            raise KeyError(f"Match dimension {match_dim} already present in DataArray.")
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1452,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            raise ValueError("No capture groups found in pattern.")
+
+        if group_dim in self._obj.dims:
+            raise KeyError(f"Group dimension {group_dim} already present in DataArray.")
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -846,68 +1452,881 @@ def replace(self, pat, repl, n=-1, case=None, flags=0, regex=True):
+            )
+
+        if dim is not None and dim in self._obj.dims:
+            raise KeyError(f"Dimension {dim} already present in DataArray.")
```
Fixed

toddrjen commented on xarray/core/accessor_str.py:
```diff
@@ -537,52 +1073,66 @@ def contains(self, pat, case=True, flags=0, regex=True):
         return self._apply(f, dtype=bool)
 
-    def strip(self, to_strip=None, side="both"):
+    def strip(
```
Fixe

toddrjen commented:
The version here should be complete, in that all planned features are implemented, although of course there may be additional changes.  So I removed the `[WIP]` part and updated `whats-new.rst` and others. I squashed my commits down and force-pushed to get a clean look at things. Please take a look and tell me what you think.

toddrjen commented:
All tests now pass as well.

andersy005 reviewed (approved):
Thank you for your contribution, @toddrjen!

max-sixty commented:
Thank you very much @toddrjen — it's a huge contribution!

max-sixty commented:
I fixed a conflict and am merging.

Thanks @toddrjen ! This is a very significant contribution.

mathause commented:
Thanks - awesome! Even going through the code took ages so kudos for sticking with it!
